graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java

```diff
@@ -258,8 +260,11 @@ public void increment() {
         }
 
         public boolean isMatched() {
-            // If a stream has multiple stream rules, all of the rules have to match.
-            return ruleCount == matches;
+            switch (stream.getMatchingType()) {
```

Comment:
Hm, but inlining the methods of a hash lookup is still slower than a constant/variable lookup. So if the method is called very often (as in: for every incoming message), we could still benefit from saving the matching type into a final variable in the constructor.